Numeric literals (binary and digit separators)

khyperia opened the pull request:
This PR implements binary literals for C# and VB, as well as digit separators (underscores) for both C# and VB. It also implements a feature check and two /features flags for binary literals and digit separators, again for both C# and VB.

This implements proposals #215 and #216

Original PR [here](https://github.com/dotnet/roslyn/pull/2730), closed due to branching reorganization.

@gafter @jaredpar @VSadov @AlekseyTs @agocke Please review

gafter commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -4599,6 +4599,12 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
+    <value>binary literals</value>
+  </data>
+  <data name="IDS_FeatureDigitSeparator" xml:space="preserve">
+    <value>numerics with underscores</value>
```
I suggest "underscores in numeric literals"

gafter replied:
But it should at least be consistent between C# and VB, which currently uses `digit separators`. Actually, I like `digit separators`.

gafter commented on src/Compilers/CSharp/Portable/Parser/Lexer.cs:
```diff
@@ -979,11 +1040,7 @@ private bool ScanNumericLiteral(ref TokenInfo info)
-                    _builder.Append(ch);
-                    TextWindow.AdvanceChar();
-                }
+                ScanNumericLiteralSingleInteger(_builder, ref underscoreInWrongPlace, isHex: false, isBinary: false);
```
Why `isBinary: false`? It looks like it should be possible to have binary here.

In fact, why isn't this line
```cs
                ScanNumericLiteralSingleInteger(_builder, ref underscoreInWrongPlace, isHex: isHex, isBinary: isBinary);
```

and hoisted out of the if-then-else?

gafter replied:
Oh, I see, hex and binary are handled in the if part.

gafter commented on src/Compilers/CSharp/Portable/Parser/Lexer.cs:
```diff
@@ -921,11 +921,65 @@ private bool ScanInteger()
+                ch = TextWindow.PeekChar();
+                if (ch == '_')
+                {
+                    CheckFeatureAvailability(MessageID.IDS_FeatureDigitSeparator);
```
This should probably be done once rather than on every `_`, as it appears in the VB implementation.

gafter commented on src/Compilers/CSharp/Portable/Parser/CharacterInfo.cs:
```diff
@@ -23,6 +23,16 @@ internal static bool IsHexDigit(char c)
+        /// <returns>true if the character is a binary digit.</returns>
+        internal static bool IsBinaryDigit(char c)
+        {
+            return c == '0' || c == '1';
```
This may be frequent, so I recommend avoiding control-flow. Try `|` instead of `||`.

gafter commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -1804,26 +1806,54 @@ FullWidthRepeat:
+                        IntegerLiteralStart = Here
+                        Base = LiteralBase.Binary
+
+                        UnderscoreInWrongPlace = Peek(Here) = "_"c
```
Do we need to call `CanGet` before calling `Peek`?

gafter replied:
And 2 other places.

gafter commented on src/Compilers/CSharp/Portable/Parser/Lexer.cs:
```diff
@@ -1221,6 +1280,27 @@ private bool ScanNumericLiteral(ref TokenInfo info)
+                }
+                if (!SyntaxFacts.IsBinaryDigit(text[i]))
+                {
+                    return false;
```
Can this occur?

AdamSpeight2008 replied:
Yes, We don't know at this stage, the function just sees characters. It isn't a binary unless all of the digit either 0 or 1, only then do now it is binary like.

gafter replied:
@AdamSpeight2008 It appears that `ScanNumericLiteralSingleInteger` won't place non-binary digits in the string. How would other characters get in there?

AdamSpeight2008 replied:
@gafter It is possible if you just consider it as pure function in isolation.
. Does it manipulate anything external to this function, excluding parameters?  
What possible values can a string have?  Is it restricted to 0 and 1 (widechar variants inc) ? No.

gafter replied:
@AdamSpeight2008 While your comment is technically correct, it is not at all helpful. This is not a pure function in isolation. It is a private instance method. My question was for the author of this PR, not a philosophical question.

gafter commented on src/Compilers/CSharp/Portable/Parser/Lexer.cs:
```diff
@@ -1221,6 +1280,27 @@ private bool ScanNumericLiteral(ref TokenInfo info)
+        private bool TryParseBinaryUInt64(string text, out ulong value)
+        {
+            value = 0;
+            for (int i = 0; i < text.Length; i++)
```
```cs
foreach (char c in text)
```

AdamSpeight2008 replied:
@gafter Slightly more efficient the way @khyperia has written since there no enumerator

gafter replied:
I believe the compiler generates the loop using an index in this case.

gafter commented:
LGTM

jaredpar commented on src/Compilers/CSharp/Portable/Errors/MessageID.cs:
```diff
@@ -149,6 +151,11 @@ internal static string RequiredFeature(this MessageID feature)
         {
```
Can we add a comment to this function about it's intent.

khyperia replied:
Seems like that should have gone in the original PR for the /features flag. I can throw it in here, but if we're merging this into a features/numeric-literals branch, we probably want that comment in future.

jaredpar replied:
@khyperia agree. This PR is against future though which is why I commented about adding it here.

jaredpar commented on src/Compilers/CSharp/Portable/Parser/Lexer.cs:
```diff
@@ -921,11 +921,65 @@ private bool ScanInteger()
+
+            char ch;
+            var lastCharWasUnderscore = false;
+            while (true)
```
Can we add a test where the text input ends in a binary literal to ensure this loop properly terminates when presented with an end of input situation?

khyperia replied:
I think this test already does so - https://github.com/khyperia/roslyn/blob/NumericLiterals/src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalTests.cs#L2351 (not sure if this link will work for posterity, but LexicalTests.cs TestNumericBinaryLiteral())

As far as I understand it, it just invokes the lexer with the specified text, with an immediate EOF afterwards.

jaredpar replied:
@khyperia you're correct, that test should cover that particular case.

khyperia replied:
I just looked into it (instead of blindly stating "it works"), TextWindow.PeekChar returns InvalidCharacter (char.MaxValue) when past the end of input, which SyntaxFacts.Is*Digit returns false for. So both test and mental tracing logic say it should work.

jaredpar commented:
Overall LGTM

Before merging though we need to add tests to ensure these new numbers formats will issue an error when used in C# 6.0 or VB 14.0.  This is an area of testing that caused us some embarassing last minute bugs recently.  Please add test coverage for this.  Essentially that parsing all of the number types when using /langversion:6 produces an error.

khyperia commented:
Ping @gafter @jaredpar Updated and responded to feedback.

@gafter re: checking IsBinaryDigit in Lexer.cs, I removed the explicit check in Lexer.cs and added a BinaryValue(char) to SyntaxFacts (before it was re-using DecValue), which has a debug assert to make sure it's a binary character (DecValue/HexValue also have a debug assert). Not sure if that's right, so pointing it out so you can re-review.

@jaredpar Added docs to RequiredFeature. I also added tests to make sure that the features need the /features flag.

agocke commented on src/Compilers/CSharp/Portable/Parser/Lexer.cs:
```diff
@@ -921,11 +921,65 @@ private bool ScanInteger()
+                    lastCharWasUnderscore = false;
+                }
+                TextWindow.AdvanceChar();
+            }
```
I would write the previous as the following:

```csharp
while (true)
{
    ch = TextWindow.PeekChar();
    if (ch == '_')
    {
        usedUnderscore = true;
        lastCharWasUnderscore = true;
    }
    else if ((isHex && !SyntaxFacts.IsHexDigit(ch))
            || (isBinary && !SyntaxFacts.IsBinaryDigit(ch))
            || (!SyntaxFacts.IsDecDigit(ch)))
    {
        break;
    }
    else
    {
        _builder.Append(ch);
        lastCharWasUnderscore = false;
    }
    TextWindow.AdvanceChar();
}
```

AdamSpeight2008 replied:
@khyperia ` II ` is only 2 addition IL bytes, vs the addition IL bytes of the additional instructions

khyperia replied:
Just tried it, I don't think the new version is equivalent (test cases failing). Example:

Say we're parsing a hex, so isHex == true, and the current character is 'a'. `(isHex && !SyntaxFacts.IsHexDigit(ch))` is false, so it moves to the next or. `isBinary` is false, so skips that too. `SyntaxFacts.IsDecDigit('a')` is false, so the `!` makes it true, so the entire expression is true and we break out of the loop incorrectly.

I do like the shortened version, but I'm not sure how to fix it to not "accidentally jump into other digit's conditions" without getting very verbose/obfuscated.

agocke replied:
You're right, I simplified too far. You should be able to add the preconditions:

```csharp
else if ((isHex && !SyntaxFacts.IsHexDigit(ch))
            || (!isHex && isBinary && !SyntaxFacts.IsBinaryDigit(ch))
            || (!isHex && !isBinary !SyntaxFacts.IsDecDigit(ch)))
    {
        break;
    }

agocke replied:
Alternatively, just split it into a bunch of else-ifs that all break.

agocke replied:
Alternatively, just split it into a bunch of else-ifs that all break.

agocke replied:
(if you validate the input you can also cut out the !isHex on the second line)

gafter replied:
What about

```cs
else if (isHex ? !SyntaxFacts.IsHexDigit(ch) :
        isBinary ? !SyntaxFacts.IsBinaryDigit(ch) :
        !SyntaxFacts.IsDecDigit(ch))
```

agocke commented on src/Compilers/CSharp/Portable/Parser/Lexer.cs:
```diff
@@ -933,23 +987,31 @@ private bool ScanNumericLiteral(ref TokenInfo info)
+                    TextWindow.AdvanceChar(2);
+                    isHex = true;
+                }
+                else if (ch == 'b' || ch == 'B')
```
So C# does not accept the full-width version of these chars and VB does?

AdamSpeight2008 replied:
@agocke  Yep [See](https://github.com/dotnet/roslyn/blob/master/src/Compilers/VisualBasic/Portable/Scanner/CharacterInfo.vb#L390)

agocke replied:
@AdamSpeight2008 I agree that's what the code implies, I'm just confirming that's intentional in the design docs.

khyperia replied:
Yes. My understanding is that C# generally disallows full-width characters for built-in things - numbers, keywords, etc. (identifiers it's okay with).

E.g. `ｏｂｊｅｃｔ ｏｂｊ = null;` causes "type not found" on `object` (it does not recognize it as a keyword), but accepts `obj` as a valid identifier.

agocke replied:
Cool.

agocke commented:
:+1:, barring comment about refactoring.

jaredpar commented:
:+1:

jaredpar commented:
@dotnet-bot test this please
